src/main/java/com/google/javascript/cl2dts/DeclarationGenerator.java

```diff
@@ -211,12 +253,28 @@ protected void printSummary() {}
 
   public String produceDts(Compiler compiler) {
     Set<String> provides = new HashSet<>();
-    // TODO: only inspect inputs which were explicitly provided by the user
+    Set<String> transitiveProvides = new HashSet<>();
+    List<String> depgraphRoots = parseDepgraphRoots();
+    out = new StringWriter();
+
+    if (depgraphRoots.isEmpty() && opts.debugOutput) {
+      emit("// WARNING: depgraph contains no roots; all provides from all inputs will be walked");
+      emitBreak();
+    }
     for (CompilerInput compilerInput : compiler.getInputsById().values()) {
-      provides.addAll(compilerInput.getProvides());
+      transitiveProvides.addAll(compilerInput.getProvides());
+      if (depgraphRoots.isEmpty() ||
+          depgraphRoots.contains(compilerInput.getSourceFile().getOriginalPath())) {
+        provides.addAll(compilerInput.getProvides());
+        if (opts.debugOutput) {
```

Comment:
Yeah, I was on the fence about google3 users seeing this, but it might help them self-diagnose problems or at least bootstrap our investigation.
Added the stripping on the actual in the assertion